Lazily initialize alias db constant prop

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54640 Lazily initialize alias db constant prop**

If we are running constant propagation on a graph that doesn't have any operators with constant inputs and any mutable inputs/outputs, we do not need to initialize an alias db. This is going to be used to speed up symbolic shape analysis.

Differential Revision: [D27340863](https://our.internmc.facebook.com/intern/diff/D27340863)

eellison committed:
Lazily initialize alias db constant prop

[ghstack-poisoned]

penguinwu commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -274,12 +274,12 @@ struct ConstantPropagator {
     }
-    // an output was removed
-    return initial_outputs != true_block->outputs().size();
+    made_change_ |= initial_outputs != true_block->outputs().size();
```
I think it is safer to track whether `replaceAndRemoveIfOutput` (e.g., make it return bool) made changes to the graph to set `made_change_`. In the future if `replaceAndRemoveIfOutput` may alter the graph without changing the number of nodes, the code would still be correct.

eellison replied:
i dont know how realistic a scenario that removeExtraIfOutputs is going to add new outputs.. but i can make the change

eellison replied:
i dont know how realistic a scenario that removeExtraIfOutputs is going to add new outputs.. but i can make the change

penguinwu commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -142,11 +143,7 @@ struct ConstantPropagator {
-    } else {
-      aliasDb_ = nullptr;
-    }
+    aliasing_types_ = aliasing_types;
```
what's the rationale behind removing `torch::make_unique`?

eellison replied:
it's to lazy initialize alias db, since constructing it can be slow. if you're optimizing a graph with all scalars you dont need to access an aliasdb to constant propagate

penguinwu commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -300,6 +300,7 @@ struct ConstantPropagator {
         loop_body->eraseOutput(loop_body_offset + i);
       }
     }
+    made_change_ |= initial_outputs != node->outputs().size();
```
Similar to above. It is better to set `made_change_ = true` when changes are made. This could be more robust if the transformation happens to alter the graph w/o changing the number of nodes.

eellison replied:
Same question, do you think it's realistic that `removeExtraLoopOutputs` would ever add more outputs ?

bertmaher replied:
I'd hope not :-).  But it also seems a bit more intuitive/safe to set `made_change_` at the exact point where a change is made.

bertmaher reviewed (approved):
lgtm!

eellison committed:
Update on "Lazily initialize alias db constant prop"


If we are running constant propagation on a graph that doesn't have any operators with constant inputs and any mutable inputs/outputs, we do not need to initialize an alias db. This is going to be used to speed up symbolic shape analysis. 

[ghstack-poisoned]

eellison committed:
Update on "Lazily initialize alias db constant prop"


If we are running constant propagation on a graph that doesn't have any operators with constant inputs and any mutable inputs/outputs, we do not need to initialize an alias db. This is going to be used to speed up symbolic shape analysis.

Differential Revision: [D27340863](https://our.internmc.facebook.com/intern/diff/D27340863)

[ghstack-poisoned]

eellison committed:
Update on "Lazily initialize alias db constant prop"


If we are running constant propagation on a graph that doesn't have any operators with constant inputs and any mutable inputs/outputs, we do not need to initialize an alias db. This is going to be used to speed up symbolic shape analysis.

Differential Revision: [D27340863](https://our.internmc.facebook.com/intern/diff/D27340863)

[ghstack-poisoned]
